### PR TITLE
Fix interpretation of integers in a BIT column

### PR DESCRIPTION
**Pull request by kennytm**
### What problem does this PR solve? 

Allows the following to pass:

```sql
CREATE TABLE t(a BIT(1));
INSERT INTO t VALUES (1);
```

### What is changed and how it works?

Previously all integers are written as a string datum, and we use `CastValue` to convert the string into the target types on demand, the result is the same for numbers. But turns out casting a string `'1'` and integer `1` into BIT type behave differently, the former will become the ASCII value instead (i.e. `x'31'` instead of `x'01'`). 

So in this PR we just add a special token for integers from the parser, and convert them into integer datum early on.

This PR does not change handling of CSV files, which always treat all fields as strings.

### Check List 

Tests 

 - Unit test
 - Integration test

Side effects

Related changes

 - Need to cherry-pick to the release branch
    * Needs to be cherry-picked before 2.1.8 release to avoid regression.

### CODE REVIEW

**Comment by kennytm** [comment]
PTAL @lonng @GregoryIan

File: lightning/mydump/loader_test.go [hunk 1]
```diff
@@ -62,8 +62,8 @@ func (s *testMydumpLoaderSuite) TestLoader(c *C) {
 	}{
 		{name: "i", dataFiles: 1},
 		{name: "report_case_high_risk", dataFiles: 1},
-		{name: "tbl_autoid", dataFiles: 1},
 		{name: "tbl_multi_index", dataFiles: 1},
+		{name: "tbl_autoid", dataFiles: 1},
```
Comment (lonng) [inline lightning/mydump/loader_test.go]: It'OK, but seems that is an unnecessary change, what does it affect?

Comment (kennytm) [reply lightning/mydump/loader_test.go]: Oh the test was broken by #156, seems `tbl_multi_index` is smaller than `tbl_autoid`.

File: lightning/mydump/parser.go [hunk 2]
```diff
@@ -154,12 +155,51 @@ const (
...
 	tokSingleQuoted
 	tokDoubleQuoted
 	tokBackQuoted
 	tokUnquoted
 )
 
+// String implements the fmt.Stringer interface
+//
+// Mainly used for debugging a token.
+func (t token) String() string {
```
Comment (lonng) [inline lightning/mydump/parser.go]: I think the better way to implement the `fmt.Stringer` here is using a string array because of the token enum is an autoincrement number. For example,
```go
var tokStringify = [...] {
	tokNil: "<Nil>",
	tokRowBegin: "RowBegin",
	// ...
}
func (t token) String() string {
	if t < tokNil || t > tokUnquoted || tokStringify[t] == "" {
		return fmt.Sprintf("unknown(%d), t)
	}
	return tokStringify[t]
}
```

File: lightning/mydump/loader_test.go (hunk 1 above)
Comment (lonng) [reply lightning/mydump/loader_test.go]: Got it.

**Comment by GregoryIan** [comment]
should we also add test cases for `enum` or `set` type?

**Comment by kennytm** [comment]
PTAL @GregoryIan 411c4766c97f0fe30cc2945c2b70a3036e19b81b

Also tried to add a test for hex literals, but turns out TiDB and MySQL are incompatible again (pingcap/tidb#10154) 🤷

**Comment by GregoryIan** [comment]
🤣 LGTM
